Analytics-CSharp/Segment/Analytics/Utilities/SyncEventPipeline.cs

```diff
@@ -0,0 +1,204 @@
+using System.Collections.Generic;
+using System.Threading;
+using global::System;
+using global::System.Linq;
+using Segment.Analytics.Policies;
+using Segment.Concurrent;
+using Segment.Serialization;
+
+namespace Segment.Analytics.Utilities
+{
+    public class SyncEventPipeline: IEventPipeline
+    {
+        private readonly Analytics _analytics;
+
+        private readonly string _logTag;
+
+        private readonly IList<IFlushPolicy> _flushPolicies;
+
+        private Channel<RawEvent> _writeChannel;
+
+        private Channel<string> _uploadChannel;
+        private Channel<string> _flushCompleteChannel;
+
+        private readonly HTTPClient _httpClient;
+
+        private readonly IStorage _storage;
+
+        internal string ApiHost { get; set; }
+
+        public bool Running { get; private set; }
+
+        internal const string FlushPoison = "#!flush";
+
+        internal static readonly RawEvent s_flushEvent = new ScreenEvent(FlushPoison, FlushPoison);
+
+        internal const string UploadSig = "#!upload";
+        internal const string FlushCompleteSig = "#!complete";
+        internal int _flushTimeout = 5000;
+        internal CancellationToken? _flushCancellationToken = null;
+
+        public SyncEventPipeline(
+            Analytics analytics,
+            string logTag,
+            string apiKey,
+            IList<IFlushPolicy> flushPolicies,
+            string apiHost = HTTPClient.DefaultAPIHost,
+            int flushTimeout = 5000,
+            CancellationToken? flushCancellationToken = null)
+        {
+            _analytics = analytics;
+            _logTag = logTag;
+            _flushPolicies = flushPolicies;
+            ApiHost = apiHost;
+
+            _writeChannel = new Channel<RawEvent>();
+            _uploadChannel = new Channel<string>();
+            _flushCompleteChannel = null;
+            _httpClient = analytics.Configuration.HttpClientProvider.CreateHTTPClient(apiKey, apiHost: apiHost);
+            _httpClient.AnalyticsRef = analytics;
+            _storage = analytics.Storage;
+            Running = false;
+            _flushTimeout = flushTimeout;
+            _flushCancellationToken = flushCancellationToken;
+        }
+
+        public void Put(RawEvent @event) => _writeChannel.Send(@event);
+
+        public void Flush() {
+            _flushCompleteChannel = new Channel<string>();
+            _writeChannel.Send(s_flushEvent);
+            _flushCompleteChannel.Receive().Wait(_flushTimeout,_flushCancellationToken??CancellationToken.None);
```

Comment:
this is a very smart way of forcing the current thread to wait for completion, but is not a very performant way and there will be issues in the situation multiple flush has been called. say there are 3 flush calls from 3 different threads, they will be all released if any of the flush request is handled, which gives inaccurate completion result. a similar way of this approach is barrier semaphore, but I don't think we need to implement it this way. simply change the implementation flush to a normal blocking method without using channel should be good enough